### PR TITLE
DMP-123 add lockProvider functionality to prevent more than 1 schedul…

### PR DESCRIPTION
**Pull request by cmagowan-hmcts**
…er running on different deployments of microservice.

**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/resources/db/migration/notification_table.sql [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 -- DROP TABLE IF EXISTS darts_schema.notification;
```
Comment (davet1985) [inline src/main/resources/db/migration/notification_table.sql:3]:
```suggestion
-- DROP TABLE IF EXISTS public.notification;
```

**Comment by cmagowan-hmcts** [comment]
Can you update the main PR comment?

File: src/main/java/uk/gov/hmcts/darts/common/config/SchedLockConfig.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Profile;
+import org.springframework.scheduling.annotation.EnableScheduling;
+
+import javax.sql.DataSource;
+
+@Configuration
+@EnableScheduling
+@EnableSchedulerLock(defaultLockAtMostFor = "PT15M")
+@Profile("!intTest")
```
Comment (davet1985) [inline src/main/java/uk/gov/hmcts/darts/common/config/SchedLockConfig.java:16]: What does this mean?

File: build.gradle [hunk 3]
```diff
@@ -99,7 +99,7 @@ task integration(type: Test) {
   group = "Verification"
   testClassesDirs = sourceSets.integrationTest.output.classesDirs
   classpath = sourceSets.integrationTest.runtimeClasspath
-  failFast = true
+  failFast = false
```
Comment (davet1985) [inline build.gradle:102]: Is this changed intended / desirable?

File: src/main/resources/db/migration/V1_5__shedlock_table.sql [hunk 4]
```diff
@@ -0,0 +1,2 @@
+CREATE TABLE shedlock(name VARCHAR(64) NOT NULL, lock_until TIMESTAMP NOT NULL,
+    locked_at TIMESTAMP NOT NULL, locked_by VARCHAR(255) NOT NULL, PRIMARY KEY (name));
```
Comment (davet1985) [inline src/main/resources/db/migration/V1_5__shedlock_table.sql:2]: Can you please format this the same as the other SQL migration files.

File: src/main/java/uk/gov/hmcts/darts/notification/service/impl/NotificationServiceImpl.java [hunk 5]
```diff
@@ -73,6 +74,8 @@ private Notification saveNotificationToDb(String eventId, String caseId, String
 
 
     @Override
+    @SchedulerLock(name = "NotificationService_sendNotificationToGovNotify",
+        lockAtLeastFor = "PT5M", lockAtMostFor = "PT14M")
```
Comment (davet1985) [inline src/main/java/uk/gov/hmcts/darts/notification/service/impl/NotificationServiceImpl.java:78]: I don't think we want to hold the lock for at least 5 minutes, maybe just 1 minute? The task runs every 2 minutes.

Also, not sure we want to hold the lock for 14 minutes? Maybe 5?

File: build.gradle (hunk 3 above)
Comment (cmagowan-hmcts) [reply build.gradle:102]: I asked Marc to change this. failFast stops any more tests from running once 1 has failed. So you you don't get the full picture of how many failing tests there are. i didn't know why we would want that.

File: src/main/java/uk/gov/hmcts/darts/common/config/SchedLockConfig.java (hunk 2 above)
Comment (cmagowan-hmcts) [reply src/main/java/uk/gov/hmcts/darts/common/config/SchedLockConfig.java:16]: This disables the SchedulerLock when the intTest profile runs. it was causing tests to fail

**Comment by davet1985** [comment]
Is this ready for a final review? Looks like it needs updating and there is a build error.

**Comment by davet1985** [comment]
@cmagowan-hmcts would you mind giving this a final review?

**Comment by cmagowan-hmcts** [comment]
Looks fine. but i cannot approve as i am a contributor

**Pull request merged**